elastic/rendezvous: make barrier and rank assignment operations O(n) instead of O(n^2)

d4l3k opened the pull request:
Summary:
This makes barrier and rank operations linear instead of quadratic with the number of workers. This drastically improves performance for rendezvous when running with over 1000 hosts.

This uses 2 approaches for different areas:

* local rank assignment: each worker does 1 set and 1 get, local ranks are assigned on the rank 0 host in a O(n) operation which reduces total store operations to be linear with number of workers.
* exit_barrier: use a counter and a final flag so each worker has to do max 1 set, 1 get and 1 add.

At 4000 hosts we see torchelastic be able to run in as little as 10 seconds down from 373 seconds.

Test Plan:
This is testing using many small tests running on a remote cluster.

{D56549942}

```
torchx run --scheduler mast -- --image=torchelastic_benchmark --j=4000x1
```

Differential Revision: D56605193


cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

kurman commented on torch/distributed/elastic/agent/server/api.py:
```diff
@@ -635,52 +616,63 @@ def _assign_worker_ranks(
         )
-        role_world_size, role_ranks = self._get_ranks(
-            role_infos, role_pos, role_start_idx, role_end_idx + 1
+        store.set(f"torchelastic/role_info/{group_rank}", agent_role_info.serialize())
```
Nit: extract prefix

d4l3k replied:
done

kurman commented on torch/distributed/elastic/agent/server/api.py:
```diff
@@ -635,52 +616,63 @@ def _assign_worker_ranks(
-            idx
-            for idx, role_info in enumerate(role_infos)
-            if _RoleInstanceInfo.compare(role_info, my_role_info) == 0
+        agent_role_info = _RoleInstanceInfo(
```
Update docs with the new shiny algo above

d4l3k replied:
done

kurman commented on torch/distributed/elastic/agent/server/api.py:
```diff
@@ -635,52 +616,63 @@ def _assign_worker_ranks(
-            role_infos, role_pos, role_start_idx, role_end_idx + 1
+        store.set(f"torchelastic/role_info/{group_rank}", agent_role_info.serialize())
+
+        # tcp store is collocated with rank 0 so we can use it to do extra compute to reduce overall # of operations.
```
I wonder if this assumption can be broken? This depends on the rendezvous handler, right?

d4l3k replied:
Even if it's not collocated it should still be much faster than before since it's way fewer operations

kurman commented on torch/distributed/elastic/utils/store.py:
```diff
@@ -58,21 +65,48 @@ def synchronize(
-    data = f"{rank}".encode()
-    synchronize(store, data, rank, world_size, key_prefix, barrier_timeout)
+
+    store.set_timeout(timedelta(seconds=barrier_timeout))
```
This changes default timeout, is that OK?

d4l3k replied:
That seems to be how these store functions operate -- I don't love it but keeping to existing patterns. We already set it in the existing synchronize call so I expect it's already at that value.

d4l3k replied:
Oh yeah -- since this is eliminating synchronize and synchronize was setting the timeout this behavior is exactly the same

kurman replied:
There will be cases when TCPStore will be shared with and used by trainer via `master_addr/master_port` args. Should we consider scoping it to just this place?

d4l3k replied:
Changed to scope it to just this with a context manager -- just a warning that this is a breaking change to previous behavior

kiukchung replied:
ideally the store API has `set(key, val, timeout)` and `get(key, timeout)` but unfortuantely that's not the case today. In the case of `TCPStore` this is a client-side timeout so even if we shared the "TCPStore" server between the agent and trainers this timeout won't get propagated down to the trainers.

kurman reviewed (commented):
Overall LGTM, except a concern for altering the timeout.

kiukchung commented on torch/distributed/elastic/agent/server/api.py:
```diff
@@ -624,63 +604,82 @@ def _assign_worker_ranks(
+        # tcp store is collocated with rank 0 so we can use it to do extra compute to reduce overall # of operations.
+        if group_rank == 0:
+            role_infos = [
+                _RoleInstanceInfo.deserialize(store.get(f"{ROLE_INFO_PREFIX}{i}"))
```
why not use `store.multi_get()` (just as you used `store.multi_set()` for writes) here?

d4l3k replied:
yeah, that's a fair point -- updated to use multi_get where possible

c-p-i-o commented on torch/distributed/elastic/agent/server/api.py:
```diff
@@ -624,63 +604,82 @@ def _assign_worker_ranks(
+            base_global_rank = 0
+            role_ranks = defaultdict(lambda: 0)
+
+            keys = []
+            values = []
```
Is there any performance to gained by pre-allocating space for `key` and `values`? I suspect the gain is minimal?

d4l3k replied:
Slightly but it's amortized O(1) in the python implementation https://stackoverflow.com/questions/5833907/repeatedly-appending-to-a-large-list-python-2-6-6

d4l3k commented on torch/distributed/elastic/agent/server/api.py:
```diff
@@ -624,63 +604,82 @@ def _assign_worker_ranks(
+        2. The rank 0 agent reads all the role_info from the store and
+           determines each agents worker ranks.
+        3. Determine the global rank: the global rank of the workers is computed
+           by cumlative sum of the local_world_size for all workers in front of it.
```
done

rsdcastro commented on torch/distributed/elastic/agent/server/api.py:
```diff
@@ -624,63 +604,85 @@ def _assign_worker_ranks(
+            store.multi_set(keys, values)
+
+        base_global_rank, global_world_size, base_role_rank, role_world_size = json.loads(
+            store.get(f"{ASSIGNED_RANKS_PREFIX}{group_rank}")
```
It may be worth adding a comment that get will block on the key being populated in TCPStore, as it may not be immediately obvious that this is the behavior.

d4l3k replied:
added

rsdcastro commented on torch/distributed/elastic/utils/store.py:
```diff
@@ -29,17 +49,20 @@ def get_all(store, rank: int, prefix: str, size: int):
         # before other processes finished `get_all` method
-        for node_rank in range(size):
-            store.get(f"{prefix}{node_rank}.FIN")
+        store.get(barrier_key)
```
I think we have a 10min timeout for this get? In the case other ranks hang for some reason, then this will exit? I haven't checked but as long as we propagate the error, this might be fine.

d4l3k replied:
yes, there's a timeout on this -- defaults to 300s or 5 minutes

kurman reviewed (approved):
let's address the release part + run E2E test with latest changes. Other than that lgtm!

d4l3k commented:
I've verified this with an e2e 32x8 gpu job that uses nccl init and an allreduce. Job succeeded across all 3 retries. Merging

d4l3k commented:
@pytorchbot merge

d4l3k commented:
@pytorchbot merge

d4l3k committed:
elastic/rendezvous: make barrier and rank assignment operations O(n) instead of O(n^2) (#124982)

Summary:
This makes barrier and rank operations linear instead of quadratic with the number of workers. This drastically improves performance for rendezvous when running with over 1000 hosts.

This uses 2 approaches for different areas:

* local rank assignment: each worker does 1 set and 1 get, local ranks are assigned on the rank 0 host in a O(n) operation which reduces total store operations to be linear with number of workers.
* exit_barrier: use a counter and a final flag so each worker has to do max 1 set, 1 get and 1 add.

At 4000 hosts we see torchelastic be able to run in as little as 10 seconds down from 373 seconds.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/124982

Test Plan:
This is testing using many small tests running on a remote cluster.

{D56549942}

```
torchx run --scheduler mast -- --image=torchelastic_benchmark --j=4000x1
```

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

Reviewed By: kurman

Differential Revision: D56605193

Pulled By: d4l3k
